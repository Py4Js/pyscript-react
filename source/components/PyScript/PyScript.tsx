import propTypes from "prop-types";
import {
  createElement,
  DetailedHTMLProps,
  HTMLAttributes,
  ReactHTML,
  WeakValidationMap,
} from "react";
import PyEnv, { PyEnvPropertiesBase } from "~components/PyEnv/PyEnv";
import ReactElementProps from "~types/ReactElementProps/ReactElementProps";

export type PyScriptPropertiesBase = Omit<
  ReactElementProps<
    DetailedHTMLProps<HTMLAttributes<HTMLElement>, HTMLElement>
  >,
  "children"
>;

export type PyScriptPropertiesWithChildren = {
  src?: never;
  children: string;
  source?: never;
};

export type PyScriptPropertiesWithSourceFullName = {
  src?: never;
  children?: never;
  source: string;
};

export type PyScriptPropertiesWithSource = {
  src: string;
  children?: never;
  source?: never;
};

export type PyScriptPropertiesWitOutputPart = {
  output: string;
  generateOutputTag?: boolean | keyof ReactHTML;
};

export type PyScriptPropertiesWitoutOutputPart = {
  output?: never;
  generateOutputTag?: never;
};

export type PyScriptPropertiesWithPyEnvPart = {
  pyEnvContent: PyEnvPropertiesBase["children"];
  pyEnvProps?: Omit<PyEnvPropertiesBase, "children">;
};

export type PyScriptPropertiesWithWithoutPyEnvPart = {
  pyEnvContent?: never;
  pyEnvProps?: never;
};

export type PyScriptPropertiesWithoutGeneric = PyScriptPropertiesBase &
  (PyScriptPropertiesWithPyEnvPart | PyScriptPropertiesWithWithoutPyEnvPart) &
  (PyScriptPropertiesWitOutputPart | PyScriptPropertiesWitoutOutputPart) &
  (
    | PyScriptPropertiesWithSource
    | PyScriptPropertiesWithChildren
    | PyScriptPropertiesWithSourceFullName
  );

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export type PyScriptProperties<T> = T extends infer T
  ? T & PyScriptPropertiesWithoutGeneric
  : PyScriptPropertiesWithoutGeneric;

export type PyScriptTag = {
  <T extends object>(properties: PyScriptProperties<T>): JSX.Element;
  propTypes: WeakValidationMap<PyScriptPropertiesWithoutGeneric>;
};

const PyScript: PyScriptTag = <T extends object>({
  children,
  output,
  generateOutputTag,
  pyEnvContent,
  pyEnvProps,
  src,
  source,
  ...rest
}: PyScriptProperties<T>): JSX.Element => {
  return (
    <>
      {pyEnvContent && <PyEnv {...pyEnvProps}>{pyEnvContent}</PyEnv>}
      {output &&
        generateOutputTag &&
        createElement(
          typeof generateOutputTag === "string" ? generateOutputTag : "div",
          {
            id: output,
          },
        )}
      <py-script {...rest} src={source || src} output={output}>
        {children || ""}
      </py-script>
    </>
  );
};

PyScript.propTypes = {
  children: propTypes.string,
  output: propTypes.string,
  generateOutputTag: propTypes.oneOfType([propTypes.string, propTypes.bool]),
  pyEnvContent: propTypes.oneOfType([propTypes.string, propTypes.array]),
  pyEnvProps: propTypes.object,
  src: propTypes.string,
  source: propTypes.string,
} as WeakValidationMap<PyScriptPropertiesWithoutGeneric>;

export default PyScript;
