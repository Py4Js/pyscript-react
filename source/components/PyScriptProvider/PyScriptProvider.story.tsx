import { ComponentMeta, ComponentStory } from "@storybook/react";
import { FunctionComponent } from "react";
import PyScriptProvider from "~components/PyScriptProvider/PyScriptProvider";
import { PyScriptProviderProperties } from "~components/PyScriptProvider/PyScriptProvider.types";

export default {
  title: "PyScriptProvider",
  component: PyScriptProvider,
} as ComponentMeta<typeof PyScriptProvider>;

const Template: ComponentStory<
  FunctionComponent<PyScriptProviderProperties>
> = ({ ...rest }: PyScriptProviderProperties): JSX.Element => {
  return (
    <PyScriptProvider {...rest}>
      {/* eslint-disable-next-line react/no-unescaped-entities*/}
      <py-script>print("Hello World!")</py-script>
    </PyScriptProvider>
  );
};

export const PyScriptProviderExample: ComponentStory<typeof PyScriptProvider> =
  Template.bind({});

PyScriptProviderExample.args = {
  cssSrc: "https://pyscript.net/alpha/pyscript.css",
  jsSrc: "https://pyscript.net/alpha/pyscript.js",
};
