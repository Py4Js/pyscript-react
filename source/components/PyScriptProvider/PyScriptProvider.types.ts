import {
  DetailedHTMLProps,
  LinkHTMLAttributes,
  PropsWithChildren,
  ScriptHTMLAttributes,
} from "react";
import { HelmetProps, ProviderProps } from "react-helmet-async";

export type PyScriptProviderPropertiesWithCSSSourceFullname = {
  cssSource?: string;
  cssSrc?: never;
};

export type PyScriptProviderPropertiesWithoutCSSSourceFullname = {
  cssSource?: never;
  cssSrc?: string;
};

export type PyScriptProviderPropertiesWithJSourceFullname = {
  jsSrc?: never;
  jsSource?: string;
};

export type PyScriptProviderPropertiesWithoutJSourceFullname = {
  jsSrc?: string;
  jsSource?: never;
};

export type PyScriptProviderPropertiesWithoutCSSPropertiesFullname = {
  cssProps?: Omit<
    DetailedHTMLProps<LinkHTMLAttributes<HTMLLinkElement>, HTMLLinkElement>,
    "href"
  >;
  cssProperties?: never;
};

export type PyScriptProviderPropertiesWithCSSPropertiesFullname = {
  cssProps?: never;
  cssProperties?: Omit<
    DetailedHTMLProps<LinkHTMLAttributes<HTMLLinkElement>, HTMLLinkElement>,
    "href"
  >;
};

export type PyScriptProviderPropertiesWithJSPropertiesFullname = {
  jsProps?: never;
  jsProperties?: Omit<
    DetailedHTMLProps<
      ScriptHTMLAttributes<HTMLScriptElement>,
      HTMLScriptElement
    >,
    "src"
  >;
};

export type PyScriptProviderPropertiesWithoutJSPropertiesFullname = {
  jsProps?: Omit<
    DetailedHTMLProps<
      ScriptHTMLAttributes<HTMLScriptElement>,
      HTMLScriptElement
    >,
    "src"
  >;
  jsProperties?: never;
};

export type PyScriptProviderPropertiesWithoutHelmetPropertiesFullname = {
  helmetProps?: HelmetProps;
  helmetProperties?: never;
};

export type PyScriptProviderPropertiesWithHelmetPropertiesFullname = {
  helmetProps?: never;
  helmetProperties?: HelmetProps;
};

export type PyScriptProviderPropertiesWithHelmetProviderPropertiesFullname = {
  helmetProviderProps?: never;
  helmetProviderProperties?: ProviderProps;
};

export type PyScriptProviderPropertiesWithoutHelmetProviderPropertiesFullname =
  {
    helmetProviderProps?: ProviderProps;
    helmetProviderProperties?: never;
  };

export type PyScriptProviderProperties = PropsWithChildren<
  (
    | PyScriptProviderPropertiesWithCSSSourceFullname
    | PyScriptProviderPropertiesWithoutCSSSourceFullname
  ) &
    (
      | PyScriptProviderPropertiesWithJSourceFullname
      | PyScriptProviderPropertiesWithoutJSourceFullname
    ) &
    (
      | PyScriptProviderPropertiesWithCSSPropertiesFullname
      | PyScriptProviderPropertiesWithoutCSSPropertiesFullname
    ) &
    (
      | PyScriptProviderPropertiesWithJSPropertiesFullname
      | PyScriptProviderPropertiesWithoutJSPropertiesFullname
    ) &
    (
      | PyScriptProviderPropertiesWithHelmetPropertiesFullname
      | PyScriptProviderPropertiesWithoutHelmetPropertiesFullname
    ) &
    (
      | PyScriptProviderPropertiesWithHelmetProviderPropertiesFullname
      | PyScriptProviderPropertiesWithoutHelmetProviderPropertiesFullname
    )
>;
