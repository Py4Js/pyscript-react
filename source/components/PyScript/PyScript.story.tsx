import { ComponentMeta, ComponentStory } from "@storybook/react";
import PyScriptProvider from "../PyScriptProvider/PyScriptProvider";
import PyScript, { PyScriptProperties, PyScriptTag } from "./PyScript";

export default {
  title: "PyScript",
  component: PyScript,
} as ComponentMeta<typeof PyScript>;

const Template: ComponentStory<PyScriptTag> = <T extends object>({
  ...rest
}: PyScriptProperties<T>): JSX.Element => {
  return (
    <PyScriptProvider>
      <PyScript {...rest} />
    </PyScriptProvider>
  );
};

export const PyScriptExample: ComponentStory<typeof PyScript> = Template.bind(
  {},
);

PyScriptExample.args = {
  src: "/folium_map.py",
  output: "folium",
  generateOutputTag: true,
  pyEnvContent: ["folium"],
};

export const PyScriptExampleFibonacci: ComponentStory<typeof PyScript> =
  Template.bind({});

PyScriptExampleFibonacci.args = {
  src: "/fibonacci.py",
};
