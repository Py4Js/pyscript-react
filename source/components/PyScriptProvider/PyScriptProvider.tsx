import propTypes from "prop-types";
import { FC, memo, WeakValidationMap } from "react";
import { Helmet, HelmetProvider } from "react-helmet-async";
import { PyScriptProviderProperties } from "~components/PyScriptProvider/PyScriptProvider.types";

const PyScriptProvider: FC<PyScriptProviderProperties> = ({
  cssSrc = "https://pyscript.net/alpha/pyscript.css",
  cssSource,
  jsSrc = "https://pyscript.net/alpha/pyscript.js",
  jsSource,
  children,
  cssProps,
  cssProperties,
  jsProperties,
  jsProps,
  helmetProps,
  helmetProperties,
  helmetProviderProps,
  helmetProviderProperties,
}: PyScriptProviderProperties): JSX.Element => {
  return (
    <HelmetProvider {...(helmetProviderProperties || helmetProviderProps)}>
      <Helmet {...(helmetProperties || helmetProps)}>
        <link
          rel="stylesheet"
          {...(cssProperties || cssProps)}
          href={cssSource || cssSrc}
        />
        <script defer {...(jsProperties || jsProps)} src={jsSource || jsSrc} />
      </Helmet>
      {children}
    </HelmetProvider>
  );
};

PyScriptProvider.propTypes = {
  cssSrc: propTypes.string,
  jsSrc: propTypes.string,
  cssSource: propTypes.string,
  jsSource: propTypes.string,
  children: propTypes.oneOfType([
    propTypes.arrayOf(propTypes.node),
    propTypes.node,
  ]).isRequired,
} as WeakValidationMap<PyScriptProviderProperties>;

export default memo(PyScriptProvider);
