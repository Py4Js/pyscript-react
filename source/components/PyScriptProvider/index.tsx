import PyScriptProvider from "~components/PyScriptProvider/PyScriptProvider";

export default PyScriptProvider;
