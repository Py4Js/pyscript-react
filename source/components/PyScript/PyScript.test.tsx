import { cleanup, render, RenderResult } from "@testing-library/react";
import { FC } from "react";
import { expectType } from "tsd";
import PyScript, { PyScriptTag } from "~components/PyScript/PyScript";
import PyScriptProvider from "~components/PyScriptProvider/PyScriptProvider";

afterEach(cleanup);

const PyScriptHelloWorldComponent: FC = (): JSX.Element => {
  return (
    <PyScriptProvider>
      {/* eslint-disable-next-line react/no-unescaped-entities*/}
      <PyScript data-testid="py-script">print("Hello world!")</PyScript>
    </PyScriptProvider>
  );
};
const PyScriptFoliumMapComponent: FC = (): JSX.Element => {
  return (
    <PyScriptProvider>
      <PyScript
        output="folium"
        generateOutputTag
        pyEnvContent={["folium"]}
        src="./test_data/folium_map.py"
      />
    </PyScriptProvider>
  );
};

describe("PyScript", (): void => {
  it("Renders component properly", async (): Promise<void> => {
    const { container }: RenderResult = render(<PyScriptHelloWorldComponent />);
    expect(container).toMatchSnapshot();
  });
  it("Compile component with Python properly", (): void => {
    const { getByTestId }: RenderResult = render(
      <PyScriptHelloWorldComponent />,
    );
    // eslint-disable-next-line testing-library/prefer-screen-queries
    const pyScript: HTMLElement = getByTestId("py-script");
    setTimeout((): void => {
      expect(pyScript).toContainEqual("Hello world!");
    }, 5_000);
  });
  it("Renders component properly with output, src, pyEnvContent and generateOutputTag", async (): Promise<void> => {
    const { container }: RenderResult = render(<PyScriptFoliumMapComponent />);
    expect(container).toMatchSnapshot();
  });
  it("Has correct type", (): void => {
    expectType<PyScriptTag>(PyScript);
  });
});
