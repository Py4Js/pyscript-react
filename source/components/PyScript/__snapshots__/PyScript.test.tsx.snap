// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PyScript Renders component properly 1`] = `
<div>
  <py-script
    data-testid="py-script"
  >
    print("Hello world!")
  </py-script>
</div>
`;

exports[`PyScript Renders component properly with output, src, pyEnvContent and generateOutputTag 1`] = `
<div>
  <py-env>
    - folium
  </py-env>
  <div
    id="folium"
  />
  <py-script
    output="folium"
    src="./test_data/folium_map.py"
  />
</div>
`;
